draft.md

```diff
@@ -0,0 +1,101 @@
+# Draft IPLD Unixfs Spec
+
+## Basic Structure
+
+A Unixfs is either a file or a directory.
+The top level IPLD object is a CBOR map with at least two fields: `type` and `data`
+and maybe a few other such as a version string or a set of flags.
+The `type` field is either `file` or `dir`.
+
+## IPLD `file`
+
+If an IPLD file is a leaf its CID type is `raw` (0x55) and has no structure.
+Otherwise its CID type is `dag-cbor` (0x71).
+The `type` field is set to `file` and the `data` field is an CBOR array.
```

Comment:
With the current `dag-cbor` and `bitswap` implementation it's quite limiting to require that this be an array of links. There's a block size limit of 2megs across the board. An array of links when serialized to CBOR can't be more than ~2,500 links before the node itself is larger than 2megs.

Unless we bake in a way to shard this we'll be limited to files that are smaller than ~5GB.

Reply:
Also, I'm a lot less concerned about files under 5GB than I am concerned with not being able to develop a smart chunker for fear that if I don't use the max blocks size the node will be too large.

I'm starting to think about developing a chunker for javascript bundles that uses the sourcemap from the bundler to chunk it into blocks built from each file. This should greatly reduce the new blocks that need to be pushed when new bundles are created, but the number of chunks will easily be greater 2,500.

Reply:
The plan is to use a CHAMP instead of an array of links, like we do today for sharded directories. I have an implementation of a CHAMP (HAMT) in ipld here: https://github.com/ipfs/go-hamt-ipld

Reply:
The JS implementation for unixfs sharding can be found at https://github.com/ipfs/js-ipfs-unixfs-engine/tree/master/src/hamt. It was built by @pgte a while ago.

Reply:
Maybe I'm missing something, but I've only seen hash map trees used for named key values, not for an ordered array. If we're using this as a replacement for `a CBOR Array` is there key semantics that need to be defined here in the spec for that? I'm only seeing the current JS hamt implementation used for sharded directories, not sharded file parts.

Reply:
@mikeal oh, the short answer is "don't include that many file parts in a single node". Look at how the importers work and structure the graph of file parts.

Reply:
> the short answer is "don't include that many file parts in a single node"

This isn't sufficient for at least 3 use cases I can think of.

* **Bundles** Tarballs, webpack and browserify output, etc. The ideal way to chunk these files is to chunk around the boundaries of the originating files so that changes to the bundle translate into a relatively small number of part changes. Tarballs that pack many small files, and pretty much any modern front-end bundle, is created from more than 2,500 original files so this puts it easily above the limit.
* **Compressed Files** Gzipped content, but especially streaming media files: ogg, mpeg, etc. You want the chunker to create blocks that respect the compression windows. This results in faster performance throughout the stack but is especially import when seeking within that content as the codec will always ask for the whole window that seekpoint is from and if it's in the middle of the block this translates into a delayed seek while the content buffers. These compression windows are configurable and sometimes the windows are very small, so a relatively small video file could be more than 2,500K parts.
* **Files Larger than 5GB** As [discussed here](https://github.com/ipld/interface-ipld-format/issues/41#issuecomment-400609254) there's currently a 2MB bitswap block limit. There's talk of supporting larger blocks but the larger the block the less efficient the transport will be.

It's fine if we just want to say that these use cases are out of scope for `unixfsv2`. Pushing them out of scope just means that we have some time to see how people solve these issues outside of `unixfsv2` and it may help us get things shipped faster. But we're probably signing up for a `unixfsv3` at some point in the future if we don't have any other way to work around this limitation.

Reply:
@mikeal look at how the importers work and structure the graph of file parts. The answer is still 'dont include that many file parts *in a single node*'. Ipfs chunks and structures things into a recursive tree, not just a single level with a flat array of links.

Reply:
Someone in Berlin mentioned that it was effectively an "array of arrays."

One thing to consider with this design, range requests don't work without loading every part of the file from the beginning to the start of the range. There's no information about the size of the individual parts so the only way to know how to seek is to load them all in serial. Really not ideal, especially for media uses cases because it makes seeking quite slow.

Reply:
The current design has range information, seeking is efficient and only has to load the required nodes for that graph traversal. I assume we would do exactly the same in V2

Reply:
I'd like to see our fix to this involves moving from `data: [chunks]` to `data: {parts: [chunks], partLengths: []}` because If we make the data attribute an object we can add other relevant information, like the type of chunking algorithm used, which would allow us to implement more efficient syncing between clients.